File-level events

cameel opened the pull request:
Closes #9737.
~Depends on #14506.~

cameel committed:
Allow event definitions at file level

nikola-matic commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -131,6 +131,9 @@ ASTPointer<SourceUnit> Parser::parse(CharStream& _charStream)
 			case Token::Function:
 				nodes.push_back(parseFunctionDefinition(true));
 				break;
+			case Token::Event:
+				nodes.push_back(parseEventDefinition());
+				break;
```
This is suspiciously simple, but then again, why wouldn't it be :)

cameel replied:
Yeah, I thought the same thing :) But in the end I could not come up with a test case that would break it so it seems that it's all that's needed. It makes sense, because it was a bit of an artificial limitation - it does not allow you to do new things in terms of code generation. I was mostly worried about naming conflicts in Natspec and ABI bit, but the hack from #13086 seems to have solved all issues there.

nikola-matic reviewed (commented):
Should probably at least mentioned that this is now possible in the [docs](https://github.com/ethereum/solidity/blob/develop/docs/contracts/events.rst?plain=1#L12).

cameel commented:
> Should probably at least mentioned that this is now possible in the [docs](https://github.com/ethereum/solidity/blob/develop/docs/contracts/events.rst?plain=1#L12).

Good point. I changed that bit to mention file-level events.

I also changed one of the examples to use such an event. Might be better to do it with one of the more relevant examples, but in all cases it seemed artificial to me and would IMO make those examples slightly worse. Currently they're all consistently using things defined within a single contract (errors, functions, types, etc.) so doing that just for events seemed forced and out of place.

nikola-matic reviewed (approved):
Given that there's not much code to review, I've looked through the test coverage, and didn't find anything missing, so an approve from me. This should nonetheless wait for @ekpyron's final approve.

ekpyron reviewed (approved):
Nothing comes to my mind that could actually go wrong with this
